pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,193 @@
+package alertmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/alertmanager/cluster"
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// state represents the Alertmanager silences and notification log internal state
+type state struct {
+	userID string
+
+	mtx    sync.Mutex
+	states map[string]cluster.State
+	logger log.Logger
+	reg    prometheus.Registerer
+
+	replicationFactor  int
+	replicateStateFunc func(context.Context, string, *clusterpb.Part) error
+
+	partialStateMergesTotal  *prometheus.CounterVec
+	partialStateMergesFailed *prometheus.CounterVec
+	stateReplicationTotal    *prometheus.CounterVec
+	stateReplicationFailed   *prometheus.CounterVec
+
+	msgc   chan *clusterpb.Part
+	stopc  chan struct{}
+	readyc chan struct{}
+}
+
+// newReplicatedStates creates a new state struct, which manages state to be replicated between alertmanagers.
+func newReplicatedStates(userID string, rf int, f func(context.Context, string, *clusterpb.Part) error, stopc chan struct{}, l log.Logger, r prometheus.Registerer) *state {
+
+	s := &state{
+		logger:             l,
+		userID:             userID,
+		replicateStateFunc: f,
+		replicationFactor:  rf,
+		states:             make(map[string]cluster.State, 2), // we use two, one for the notifications and one for silences.
+		msgc:               make(chan *clusterpb.Part),
+		readyc:             make(chan struct{}),
+		stopc:              stopc,
+		reg:                r,
+		partialStateMergesTotal: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_partial_state_merges_total",
+			Help: "Number of times we have received a partial state to merge for a key.",
+		}, []string{"key"}),
```

Comment:
Do we need per-key metrics or would per-user (i.e. per-state) be enough?

Reply:
per-key includes both user and state so it would always be at most 2 per user.